File: NetDeck/UIAlert+NetDeck.m
//
//  UIAlert+NetDeck.m
//  Net Deck
//
//  Created by Gereon Steffens on 28.10.14.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

#import "UIAlert+NetDeck.h"

@implementation UIAlertAction (NetDeck)

+(UIAlertAction*) cancelAction:(void (^)(UIAlertAction *action))handler
{
    NSString* title = IS_IPAD ? @"" : l10n(@"Cancel");
    return [UIAlertAction actionWithTitle:title style:UIAlertActionStyleCancel handler:handler];
}

+(UIAlertAction*) cancelAlertAction:(void (^)(UIAlertAction *action))handler
{
    NSString* title = l10n(@"Cancel");
    return [UIAlertAction actionWithTitle:title style:UIAlertActionStyleCancel handler:handler];
}

+(UIAlertAction*) actionWithTitle:(NSString *)title handler:(void (^)(UIAlertAction *))handler
{
    return [UIAlertAction actionWithTitle:title style:UIAlertActionStyleDefault handler:handler];
}

@end

// see http://stackoverflow.com/questions/26554894/how-to-present-uialertcontroller-when-not-in-a-view-controller
@implementation UIAlertController(NetDeck)

+(void) alertWithTitle:(NSString*)title message:(NSString*)msg button:(NSString*)button {
    UIAlertController* alert = [UIAlertController alertWithTitle:title message:msg];
    [alert addAction:[UIAlertAction actionWithTitle:button handler:nil]];
    [alert show];
}

+(UIAlertController*) alertWithTitle:(NSString*)title message:(NSString*)msg {
    return [UIAlertController alertControllerWithTitle:title message:msg preferredStyle:UIAlertControllerStyleAlert];
}

+(UIAlertController*) actionSheetWithTitle:(NSString*)title message:(NSString*)msg {
    return [UIAlertController alertControllerWithTitle:title message:msg preferredStyle:UIAlertControllerStyleActionSheet];
}

-(void) show {
    [self present:NO completion:nil];
}

-(void) present:(BOOL)animated completion:(void (^ __nullable)(void))completion {
    UIViewController* rootVC = [UIApplication sharedApplication].keyWindow.rootViewController;
    [self presentFromController:rootVC animated:animated completion:completion];
}

-(void) presentFromController:(UIViewController*)controller animated:(BOOL)animated completion:(void (^ __nullable)(void))completion {
    if ([controller isKindOfClass:[UINavigationController class]]) {
        UINavigationController* navVC = (UINavigationController*)controller;
        if ([navVC visibleViewController]) {
            [self presentFromController:navVC.visibleViewController animated:animated completion:completion];
        }
    } else if ([controller isKindOfClass:[UITabBarController class]]) {
        UITabBarController* tabVC = (UITabBarController*)controller;
        UIViewController* selected = [tabVC selectedViewController];
        if (selected) {
            [self presentFromController:selected animated:animated completion:completion];
        }
    } else {
        [controller presentViewController:self animated:animated completion:completion];
    }
}

@end

File: NetDeck/UIAlert+NetDeck.swift
//
//  UIAlert+NetDeck.swift
//  NetDeck
//
//  Created by Gereon Steffens on 23.12.16.
//  Copyright © 2017 Gereon Steffens. All rights reserved.
//

import UIKit

extension UIAlertAction {
    typealias ActionHandler = (UIAlertAction) -> Void
    
    static func actionSheetCancel(_ handler: ActionHandler?) -> UIAlertAction {
        let title = Device.isIpad ? "" : "Cancel".localized()
        return UIAlertAction(title: title, style: .cancel, handler: handler)
    }
    
    static func alertCancel(_ handler: ActionHandler?) -> UIAlertAction {
        return UIAlertAction(title: "Cancel".localized(), style: .cancel, handler: handler)
    }
    
    static func action(title: String, handler: ActionHandler?) -> UIAlertAction {
        return UIAlertAction(title: title, style: .default, handler: handler)
    }
    
    convenience init(title: String, handler: ActionHandler?) {
        self.init(title: title, style: .default, handler: handler)
    }
}

extension UIAlertController {
    static func alert(title: String? = nil, message: String? = nil) -> UIAlertController {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        return alert
    }
    
    static func actionSheet(title: String?, message: String?) -> UIAlertController {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .actionSheet)
        return alert
    }
    
    static func alert(withTitle title: String?, message: String? = nil, button: String) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: button, style: .default, handler: nil))
        alert.show()
    }
    
    func show() {
        self.present(animated: false, completion: nil)
    }
    
    private typealias CompletionHandler = () -> Void
    
    private func present(animated: Bool, completion: CompletionHandler?) {
        guard let root = UIApplication.shared.keyWindow?.rootViewController else { return }
        self.presentFromController(root, animated: animated, completion: completion)
    }
    
    private func presentFromController(_ controller: UIViewController, animated: Bool, completion: CompletionHandler?) {
        if let nav = controller as? UINavigationController {
            if let visible = nav.visibleViewController {
                self.presentFromController(visible, animated: animated, completion: completion)
            }
        } else if let tab = controller as? UITabBarController {
            if let selected = tab.selectedViewController {
                self.presentFromController(selected, animated:  animated, completion: completion)
            }
        } else {
            controller.present(self, animated: animated, completion: completion)
        }
    }
}
